pacsman/utils.py

```diff
@@ -9,6 +9,24 @@
 from pydicom.errors import InvalidDicomError
 
 
+def process_and_write_png_from_file(thumbnail_dcm_path):
+    '''
+    :param thumbnail_dcm_path: DICOM instance file path. Must be unique per instance.
+        The file is deleted to make the PNG.
+    :return path to png (or None on failure). Uses the same name as the input file.
+    '''
+    if not os.path.exists(thumbnail_dcm_path):
+        return None
+    png_path = None
+    try:
+        thumbnail_ds = dcmread(thumbnail_dcm_path)
+        png_path = os.path.splitext(thumbnail_dcm_path)[0] + '.png'
+        process_and_write_png(thumbnail_ds, png_path)
+    finally:
+        os.remove(thumbnail_dcm_path)
+    return png_path
```

Comment:
It seems like there is still the possibility for multiple concurrent thumbnail requests to conflict with one another by jostling for the same file name. Would it make sense to implement something like a file lock in this context? I am thinking of a scenario such as somebody rapidly refreshing a page which is loading a collection of thumbnails.

In particular, I am concerned that the `thumbnail_dcm_path` could be deleted between lines 18 and 22. If we handle the downstream effects of getting `None` back as a png_path, it seems like it's not a problem.

Alternatively, perhaps you can append a random hash to the thumbnail name that is passed into this function. That way, the file name is unique for each process; then you can remove the random hash before writing the PNG. Or you could use Python's [temp file module](https://docs.python.org/3/library/tempfile.html)  for the dicom file and then pass the desired PNG file name into this function as an argument?

Reply:
This is essentially what was happening with the lymphoma patient and the filesystem client. This is because it just copied the file without changing its name, and two series had the same middle slice name. 

I think the temp file is a good idea, though. Will add this. They are used a lot in the dcmtk client.

Reply:
Thinking about it, not using a tempfile for the DICOM file does give you the chance to debug the PNG writing if it fails for some reason.

Reply:
We've only seen this happen with the filesystem client, as the instance IDs are supposed to be globally unique.

Reply:
That's a good point about being able to debug the image-- I think that's a compelling enough reason to avoid using tempfiles. I know you can do some stuff with `NamedTemporaryFiles` where you can validate that the files exist somewhere in the filesystem and you can get a valid file path name for them, but it may be more work than it's worth if the instance ID works for most cases.

Reply:
I will keep as is since we haven't seen any issues with pynetdicom client in practice. When this used to happen you'd get a "missing image" placeholder on select timepoint, so the `None` case isn't causing crashes downstream.

Reply:
Sounds good to me!